File: 17. Letter Combinations of a Phone Number/main.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        result = []

        if not digits:
            return result

        maps = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        def backtrack(index: int, combine: List[str]):
            if index == len(digits):
                result.append("".join(combine))
                return

            for c in maps[digits[index]]:
                combine.append(c)
                backtrack(index + 1, combine)
                combine.pop()

        backtrack(0, [])

        return result


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        result = []

        if not digits:
            return result

        maps = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        def backtrack(index: int, combine: str):
            if index == len(digits):
                result.append(combine)
                return

            for c in maps[digits[index]]:
                backtrack(index + 1, combine + c)

        backtrack(0, "")

        return result

File: 17. Letter Combinations of a Phone Number/main.rs
struct Solution;

use std::collections::HashMap;

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let mut result = Vec::new();

        if digits.is_empty() {
            return result;
        }

        let mut array = String::new();
        Self::backtrack(digits.as_bytes(), 0, &mut array, &mut result);

        result
    }

    fn backtrack(digits: &[u8], index: usize, array: &mut String, result: &mut Vec<String>) {
        if index == digits.len() {
            result.push(array.to_string());
            return;
        }

        let maps: HashMap<char, &str> = vec![
            ('2', "abc"),
            ('3', "def"),
            ('4', "ghi"),
            ('5', "jkl"),
            ('6', "mno"),
            ('7', "pqrs"),
            ('8', "tuv"),
            ('9', "wxyz"),
        ]
        .into_iter()
        .collect();

        for c in maps[&(digits[index] as char)].bytes() {
            array.push(c as char);
            Self::backtrack(digits, index + 1, array, result);
            array.pop();
        }
    }
}
